airbase/download_api.py

```diff
@@ -0,0 +1,544 @@
+        for poll in payload:
+            key, val = poll["notation"], pollutant_id_from_url(poll["id"])
+            ids[key].add(val)
+        return ids
+
+    @aiocache.cached()
+    async def countries(self) -> list[str]:
+        """
+        Get list of all countries
+        """
+        payload = await self.raw_country_json()
+        return [country["countryCode"] for country in payload]
+
+    @aiocache.cached()
+    async def cities(self, *countries) -> defaultdict[str, set[str]]:
+        """
+        Get list of cities in each country
+        """
+        if not COUNTRY_CODES.issuperset(countries):
+            unknown = sorted(set(countries) - COUNTRY_CODES)
+            warn(
+                f"Unknown country code(s) {', '.join(unknown)}",
+                UserWarning,
+                stacklevel=2,
+            )
+
+        payload = await self.raw_city_json(*countries)
+        cities: defaultdict[str, set[str]] = defaultdict(set)
+        for city in payload:
+            key, val = city["countryCode"], city["cityName"]
+            cities[key].add(val)
+
+        return cities
+
+    async def _download_url_to_directory(
+        self, url: str, destination: str | Path
+    ):
+        """
+        Download the content of a list of URLs ot files in a directory
+        """
+        destination = Path(destination)
+        response = await self.get(url)
+        url_path = parse.urlparse(url).path
+        filename = Path(url_path).name
+        async with aiofiles.open(destination / filename, "wb") as f:
+            await f.write(response.content)
+
+    @aiocache.cached()
+    async def _pollutant_notation_to_properties(
+        self, pollutant_notation: str
+    ) -> list[str]:
+        pollutants = await self.pollutants()
+        ids = pollutants[pollutant_notation]
+        return [
+            f"http://dd.eionet.europa.eu/vocabulary/aq/pollutant/{id}"
+            for id in ids
+        ]
+
+    async def download_downloadinfo(
+        self, *urls: DownloadInfo, destination: str | Path
+    ):
+        download_jobs = []
+
+        pbar = tqdm(
+            leave=True,
+            disable=not self.progress,
+            desc="Downloading parquets",
+            unit="files",
+            position=1,
+        )
+        pbar.set_postfix_str("Checking total number of files..")
+
+        async for parquet_url in self.parquet_file_urls(*urls):
+            task = asyncio.create_task(
+                self._download_url_to_directory(parquet_url, destination)
+            )
+
+            def update_pbar(future):
+                pbar.update(1)
+
+            task.add_done_callback(update_pbar)
+            download_jobs.append(task)
+
+        if self.progress:
+            total_files_task = asyncio.create_task(
+                self.total_num_parquet_files(*urls)
+            )
+
+            def update_pbar_total(future):
+                pbar.total = future.result()
+                pbar.set_postfix_str("")
```

Comment:
This isn't totally necessary and wastes a few requests but it's nice to have to know roughly how long the download will take